Integrate with WebDriver and WebDriver BiDi for automation

nondebug opened the pull request:
Closes #???

The following tasks have been completed:

 * [ ] Modified Web platform tests (link to pull request)

Implementation commitment:

 * [ ] [WebKit](https://bugs.webkit.org/show_bug.cgi?id=309849)
 * [ ] Chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=)
 * [ ] Gecko (https://bugzilla.mozilla.org/show_bug.cgi?id=)


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/gamepad/pull/224.html" title="Last updated on Mar 12, 2026, 11:43 PM UTC (02c8a13)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/gamepad/224/fbde104...02c8a13.html" title="Last updated on Mar 12, 2026, 11:43 PM UTC (02c8a13)">Diff</a>

nondebug committed:
Integrate with WebDriver and WebDriver-BiDi for automation

nondebug committed:
Add a model section

nondebug committed:
Split model sections

nondebug committed:
Add a section on input values and define logical values

marcoscaceres commented on index.html:
```diff
@@ -2406,6 +2608,913 @@ <h2>
+              <li>
+                Let |simulatedGamepads| be the [=simulated gamepad map=].
+              </li>
+              <li>
```
```suggestion
              <li data-cite="webcrypto">
```

nondebug committed:
Align with Gamepad API testing doc

Update webdriver CDDL type definitions.
Move all bounds information into SimulateGamepadParams.
Remove vendor/product/name parameters.
Remove "flush" action and move its functionality to "pause".
Add steps for processing and dispatching actions.

nondebug committed:
Update index.html

Co-authored-by: Marcos Cáceres <marcos@marcosc.com>

sadym-chromium commented on index.html:
```diff
@@ -136,6 +136,204 @@ <h2>
+        </p>
+        <p>
+          An [=input value=] may have an associated <dfn>HID usage identifier</dfn>, a 32-bit value that identifies the type of data represented by the input value.
+          A HID usage does not precisely describe the [=input control layout=], but by convention many [=gamepads=] with similar layouts use similar usages.
```
> by convention many [=gamepads=] with similar layouts use similar usages

It would be great to link a resource with the list of conventional HID usage identifiers. Even if it is not strictly standardizes, it would be helpful to have such a list for reference.

lauromoura replied:
For a complete list of usage identifiers, there's https://www.usb.org/document-library/hid-usage-tables-16 (a 443-page PDF)?

sadym-chromium commented on index.html:
```diff
@@ -136,6 +136,204 @@ <h2>
+        time. A [=touch point=] has a <dfn>touch x coordinate</dfn>
+        and a <dfn>touch y coordinate</dfn> representing the position in the [=touch surface=]'s coordinate system.
+        If a [=touch surface=] is on the top, bottom, front, or back side of the [=gamepad=] then the [=touch x coordinate=] is measured along the left-right axis, otherwise it is measured along the top-bottom axis.
+        The [=touch y coordinate=] is measured along the perpendicular axis.
```
NIT: should we define the specific orientation of the the perpendicular axis?

sadym-chromium commented on index.html:
```diff
@@ -587,6 +795,16 @@ <h3>
+            Return 2 (|value| − |minimum|) / (|maximum| − |minimum|) − 1.
+          </li>
+          <li>
+          </li>
```
```suggestion
```

sadym-chromium commented on index.html:
```diff
@@ -587,6 +795,16 @@ <h3>
+        </p>
+        <ol>
+          <li>
+            Return 2 (|value| − |minimum|) / (|maximum| − |minimum|) − 1.
```
I would expect the normalized output of input with min `1`, max `5` and val `3` to be `1/2`:
```suggestion
            Return (|value| − |minimum|) / (|maximum| − |minimum|).
```

sadym-chromium commented on index.html:
```diff
@@ -587,6 +795,16 @@ <h3>
+          </li>
+        </ol>
         <p>
           To <dfn>map and normalize axes</dfn> for |gamepad:Gamepad|, run the
           following steps:
```
NIT: The definition below now can be formulated in strict terms introduced previously: `input values` #dfn-input-values and `axis list` #dfn-axis-list
>
```
<li>Let |axisValues:list| be a [=list=] of {{unsigned long}} values
representing the most recent logical axis input values for each axis
input of the device represented by |gamepad|.
</li>```

sadym-chromium replied:
Also `mappedIndex` is not used, so probably it can be omitted.

sadym-chromium commented on index.html:
```diff
@@ -1361,6 +1585,8 @@ <h2>
                 <li>[=Issue a haptic effect=] to the actuator with |type|,
                 |params|, and the |playEffectTimestamp|.
                 </li>
+                <li>If |this|'s [=gamepad=] is a [=simulated gamepad=], run the steps to [=vibrate a simulated gamepad=] for |this|'s [=gamepad=] with |params|.
+                </li>
```
I assume we should do only one of these steps:
```suggestion
                <li>If |this|'s [=gamepad=] is a [=simulated gamepad=], run the steps to [=vibrate a simulated gamepad=] for |this|'s [=gamepad=] with |params|.
                </li>
                <li>Otherwise, [=issue a haptic effect=] to the actuator with |type|,
                |params|, and the |playEffectTimestamp|.
                </li>
```

sadym-chromium commented on index.html:
```diff
@@ -2255,7 +2483,7 @@ <h3 id="event-gamepadconnected">
       </h3>
       <p>
-        When a gamepad becomes available on the system, run the following
+        <dfn>When a gamepad becomes available on the system</dfn>, run the following
```
I believe it's more correct to split this into a trigger and a process.
```suggestion
        When a [=gamepad=] <var>gamepad</var> becomes available on the system,
        [=emit `gamepadconnected` event=] for <var>gamepad</var>.
      </p>
      <p>
        To <dfn>emit `gamepadconnected` event</dfn> for given [=gamepad=] <var>gamepad</var>, run the following
```

sadym-chromium replied:
Also I believe the `[=current global object=]` is ambiguous, especially in case of simulated devices.

sadym-chromium commented on index.html:
```diff
@@ -2310,7 +2538,7 @@ <h3 id="event-gamepaddisconnected">
       </h3>
       <p>
-        When a gamepad becomes unavailable on the system, run the following
+        <dfn>When a gamepad becomes unavailable on the system</dfn>, run the following
```
The same issue as with `When a gamepad becomes available on the system`.

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+        It is OPTIONAL for a [=user agent=] to support them.
+      </p>
+      <p>
+        Each [=top-level traversable=] has a <dfn>simulated gamepad map</dfn> which is an [=ordered map=] of [=version 4 UUIDs=] to [=gamepad input sources=].
```
I think the UUID key it sufficient and we don't need to have it ordered map.
```suggestion
        Each [=top-level traversable=] has a <dfn>simulated gamepad map</dfn> which is a [=map=] of [=version 4 UUID=] and [=gamepad input source=].
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+        It is OPTIONAL for a [=user agent=] to support them.
+      </p>
+      <p>
+        Each [=top-level traversable=] has a <dfn>simulated gamepad map</dfn> which is an [=ordered map=] of [=version 4 UUIDs=] to [=gamepad input sources=].
```
Perhaps it's `simulated gamepad`, not `gamepad input source`:
```suggestion
        Each [=top-level traversable=] has a <dfn>simulated gamepad map</dfn> which is an [=ordered map=] of [=version 4 UUIDs=] to [=simulated gamepad=].
```

sadym-chromium replied:
I believe the UUID version is important only when it is generated:
```suggestion
        Each [=top-level traversable=] has a <dfn>simulated gamepad map</dfn> which is an [=ordered map=] of [=UUID=] to [=gamepad input sources=].
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            <pre class="cddl">
+              emulation.GamepadBounds = {
+                max: float64,
+                ? min: float64,
```
You can define `.default` in CDDL like this:
```suggestion
                ? min: float64 .default 0.0,
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            <pre class="cddl">
+              emulation.SimulateGamepadParams = {
+                type: "gamepad",
+                context: browsingContext.BrowsingContext,
```
IDK if we can cross-reference CDDL to BiDi spec, so let's use `text` the same way [Bluetooth spec does](https://webbluetoothcg.github.io/web-bluetooth/#issue-194ad560).
```suggestion
                context: text,
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                Let |simulatedGamepads| be the [=simulated gamepad map=].
+              </li>
+              <li>
+                If |simulatedGamepads|[|params|.|deviceId|] does not [=map/exist=], abort these steps.
```
1. We should return a error if there is no such an simulated device.
2. `params` is a map, so `|params|["<code>deviceId</code>"]` is the correct way to access it's values by keys.
```suggestion
                If |simulatedGamepads|[|params|["<code>deviceId</code>"]] does not [=map/exist=], return [=error=]
                with [=error code=] [=invalid argument=].
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                If |simulatedGamepads|[|params|.|deviceId|] does not [=map/exist=], abort these steps.
+              </li>
+              <li>
+                Let |gamepad| be |simulatedGamepads|[|params|.|deviceId|].
```
```suggestion
                Let |gamepad| be |simulatedGamepads|[|params|["<code>deviceId</code>"]].
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                Run the steps [=when a gamepad becomes unavailable on the system=] with |gamepad|.
+              </li>
+              <li>
+                [=map/Remove=] |simulatedGamepads|[|params|.|deviceId|].
```
```suggestion
                [=map/Remove=] |simulatedGamepads|[|params|["<code>deviceId</code>"]].
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+              }
+            </pre>
+            <p>
+              To get <dfn>the identifier for a simulated gamepad</dfn> |gamepad|:
```
Optionally, we can also extend the definition of "simulated gamepad" as "Simulated gamepad has an associated [=UUID=] <dfn>simulated gamepad identifier</dfn>".

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                Let |deviceId:string| be [=the identifier for a simulated gamepad=] |gamepad|.
+              </li>
+              <li>
+                If |deviceId| is `null`, abort these steps.
```
Maybe we want to notify the user that the given gamepad is removed by throwing an exception?

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                Let |channels| be the [=list=] « `"strongMagnitude"`, `"weakMagnitude"`, `"leftTrigger"`, `"rightTrigger"` ».
+              </li>
+              <li>
+                [=list/For each=] |channel| of |channels|:
```
It means that we ignore all other potential values. IDK if this is correct, and how extensible it is.

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+                Let |body| be a [=map=] matching the `emulation.SimulatedDeviceEvent` production with the `params` field set to |eventParams|.
+              </li>
+              <li>
+                Let |relatedNavigables| be a [=set=] containing |navigable|.
```
Note: this means that even if the event was initiated from an iframe, the event will be emitted only on the top-level navigable. I believe this is fine.

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            Gamepad Input Source
+          </h4>
+          <p>
+            A <dfn>gamepad input source</dfn> is an [=input source=] that is associated with a [=gamepad=]-like device.
```
Do we need the `gamepad input source` instance? Cannot it be a `simulated gamepad`?

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            </dl>
+          </section>
+          <p>
+            Extend the dispatch action algorithm table in [=dispatch tick actions=] with the following table:
```
Please add reference to the WebDriver Classic spec, where `dispatch tick actions` is defined:
```
spec: WEBDRIVER; urlPrefix: https://w3c.github.io/webdriver/
    text: dispatch tick actions; url: dfn-dispatch-tick-actions
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+              Let |simulated gamepads| be the [=simulated gamepad map=].
+            </li>
+            <li>
+              If |simulated gamepads|[|device id|] does not [=map/exist=], abort these steps.
```
Return a error with "invalid argument" or a new error like "no such simulated gamepad"

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+          <p>
+            A [=gamepad input source=] has the following items:
+          </p>
+          <table>
```
Consider defining the `gamepad input source` (or `simulated gamepad`) as a [=struct=] with [=struct/item=] ([example](https://github.com/w3c/webdriver-bidi/blob/32892f339e3a749e7cf6640fe30f0ef8857a34c4/index.bs#L2955-L2957)). That would allow for referencing the items with the links to it's definitions and search for references ([example](https://www.w3.org/TR/webdriver-bidi/#download-behavior-struct-destination-folder)).

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+              Let |action| be an [=action object=] constructed with arguments |id|, `"gamepad"`, and |subtype|.
+            </li>
+            <li>
+              If |subtype| is `"pause"`, let |result| be the result of [=trying=] to [=process a pause action=] with arguments |action item| and |action|, and return |result|.
```
This and other calls can be simplified:
```suggestion
              If |subtype| is `"pause"`, return the result of [=trying=] to [=process a pause action=] |action item| and |action|.
```

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            <li>
+              If |subtype| is `"touchEnd"`, let |result| be the result of [=trying=] to [=process a gamepad touch end action=] with arguments |action item| and |action|, and return |result|.
+            </li>
+            <li>
+              Return [=success=] with data |action|.
+            </li>
```
Should not it be an "invalid argument" error, as the actions is unknown?

sadym-chromium commented on index.html:
```diff
@@ -2406,6 +2634,1425 @@ <h2>
+            Actions
+          </h4>
+          <p>
+            To <dfn>process a gamepad action</dfn> given |id| and |action item|:
```
It looks like this part is not finished yet. Please let me know if I can help with it.

nondebug committed:
Update index.html

Co-authored-by: Maksim Sadym <69349599+sadym-chromium@users.noreply.github.com>
